File: voice/backend/app/models/learning_challenge.py
"""LearningChallenge SQLAlchemy model — mirrors the Go GORM model in gather_rpg DB."""

import uuid
from datetime import datetime
from sqlalchemy import Column, String, Text, Integer, Boolean, DateTime
from sqlalchemy.dialects.postgresql import UUID as PGUUID, ARRAY
from sqlalchemy.orm import relationship

from app.db.database import Base


class LearningChallenge(Base):
    __tablename__ = "learning_challenges"

    id = Column(PGUUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    type = Column(String(30), nullable=False)
    question = Column(Text, nullable=False)
    option1 = Column(Text, nullable=False)
    option2 = Column(Text, nullable=False)
    option3 = Column(Text, nullable=False)
    correct_option = Column(Integer, nullable=False)
    explanation_es = Column(Text)
    question_es = Column(Text)
    tags = Column(ARRAY(String), nullable=False, default=[])
    difficulty = Column(String(20), nullable=False, default="beginner")
    language_learning = Column(String(20), nullable=False, default="english")
    phonetic = Column(String(100), nullable=True)
    requires_audio = Column(Boolean, nullable=False, default=False)
    audio_url = Column(String(500), nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    attempts = relationship("UserChallengeAttempt", back_populates="challenge")

File: gather-rpg-backend/internal/models/learning_challenge.go
package models

import (
	"time"

	"github.com/google/uuid"
	"github.com/lib/pq"
	"gorm.io/gorm"
)

// ChallengeType represents the type of learning challenge
type ChallengeType string

const (
	ChallengeTypeVocabulary    ChallengeType = "vocabulary"
	ChallengeTypeGrammar       ChallengeType = "grammar"
	ChallengeTypePronunciation ChallengeType = "pronunciation"
	ChallengeTypeListening     ChallengeType = "listening"
)

// DifficultyLevel represents the difficulty of a challenge
type DifficultyLevel string

const (
	DifficultyBeginner     DifficultyLevel = "beginner"
	DifficultyIntermediate DifficultyLevel = "intermediate"
	DifficultyAdvanced     DifficultyLevel = "advanced"
)

// LearningChallenge represents an English learning challenge/question
type LearningChallenge struct {
	ID               uuid.UUID       `gorm:"type:uuid;default:gen_random_uuid();primaryKey" json:"id"`
	Type             ChallengeType   `gorm:"type:varchar(30);not null" json:"type"`
	Question         string          `gorm:"type:text;not null" json:"question"`
	Option1          string          `gorm:"type:text;not null" json:"option_1"`
	Option2          string          `gorm:"type:text;not null" json:"option_2"`
	Option3          string          `gorm:"type:text;not null" json:"option_3"`
	CorrectOption    int             `gorm:"not null" json:"correct_option"` // 1, 2 or 3
	ExplanationES    string          `gorm:"type:text" json:"explanation_es"`
	QuestionES       string          `gorm:"type:text" json:"question_es"`
	Tags             pq.StringArray  `gorm:"type:text[];not null;default:'{}'" json:"tags"`
	Difficulty       DifficultyLevel `gorm:"type:varchar(20);not null;default:'beginner'" json:"difficulty"`
	LanguageLearning string          `gorm:"type:varchar(20);not null;default:'english'" json:"language_learning"`
	Phonetic         string          `gorm:"type:varchar(100)" json:"phonetic"`
	RequiresAudio    bool            `gorm:"not null;default:false" json:"requires_audio"`
	AudioURL         string          `gorm:"type:varchar(500)" json:"audio_url,omitempty"`
	CreatedAt        time.Time       `json:"created_at"`
	UpdatedAt        time.Time       `json:"updated_at"`
	DeletedAt        gorm.DeletedAt  `gorm:"index" json:"-"`

	// Relationships
	Attempts []UserChallengeAttempt `gorm:"foreignKey:ChallengeID" json:"-"`
}
